sm/json_sm/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+            "currentSoftwareList": []
+        }"#;
+
+        let actual_json = response.to_json().expect("Failed to serialize");
+        assert_eq!(actual_json, remove_whitespace(expected_json));
+    }
+
+    #[test]
+    fn using_a_software_list_error() {
+        let json_response = r#"{
+            "id": 123,
+            "status": "failed",
+            "reason": "Request timed-out"
+        }"#;
+        let response = SoftwareListResponse::from_json(json_response).expect("Failed to deserialize");
+
+        assert_eq!(response.id(), 123);
+        assert_eq!(response.error(), Some("Request timed-out".into()));
+    }
+
+    #[test]
+    fn creating_a_software_update_request() {
+        let mut request = SoftwareUpdateRequest::new(123);
+
+        request.add_updates("debian", vec![
+            SoftwareModuleUpdate::install(SoftwareModule { name: "nodered".to_string(), version: Some("1.0.0".to_string()), url: None }),
+            SoftwareModuleUpdate::install(SoftwareModule { name: "collectd".to_string(), version: Some("5.7".to_string()), url: Some("https://collectd.org/download/collectd-tarballs/collectd-5.12.0.tar.bz2".to_string()) }),
+        ]);
+
+        request.add_updates("docker", vec![
+            SoftwareModuleUpdate::install(SoftwareModule { name: "nginx".to_string(), version: Some("1.21.0".to_string()), url: None }),
+            SoftwareModuleUpdate::remove(SoftwareModule { name: "mongodb".to_string(), version: Some("4.4.6".to_string()), url: None }),
+        ]);
+
+        let expected_json = r#"{
+            "id": 123,
+            "updateList": [
+                {
+                    "type": "debian",
+                    "modules": [
+                        {
+                            "name": "nodered",
+                            "version": "1.0.0",
+                            "action": "install"
+                        },
+                        {
+                            "name": "collectd",
+                            "version": "5.7",
+                            "url": "https://collectd.org/download/collectd-tarballs/collectd-5.12.0.tar.bz2",
+                            "action": "install"
+                        }
+                    ]
+                },
+                {
+                    "type": "docker",
+                    "modules": [
+                        {
+                            "name": "nginx",
+                            "version": "1.21.0",
+                            "action": "install"
+                        },
+                        {
+                            "name": "mongodb",
+                            "version": "4.4.6",
+                            "action": "remove"
+                        }
+                    ]
+                }
+            ]
+        }"#;
+        let actual_json = request.to_json().expect("Failed to serialize");
+        assert_eq!(actual_json, remove_whitespace(expected_json));
```

Comment:
I believe you don't need this function `remove_whitespace`, you could use `split_whitespace`, something like:
```suggestion
        assert_eq!(actual_json, expected_json.split_whitespace().collect::<String>());
```